File: backend/routes/ai_routes.py
from fastapi import APIRouter, HTTPException, Request, Depends
import requests
import logging
from slowapi import Limiter
from slowapi.util import get_remote_address

router = APIRouter()

AI_BACKEND_URL = "http://127.0.0.1:8000"  # URL locale del backend AI

# ✅ Configurazione del logging avanzato
logging.basicConfig(filename="logs/ai_errors.log", level=logging.ERROR, format="%(asctime)s - %(levelname)s - %(message)s")

# ✅ Rate Limiting (Max 20 richieste per IP ogni 10 minuti)
limiter = Limiter(key_func=get_remote_address)

@router.post("/ai/chat")
@limiter.limit("20/minute")  # ✅ Limitiamo a 20 richieste al minuto per evitare abusi
async def chat_with_ai(user_message: str, session_id: str):
    """Invia un messaggio all'AI e ottiene una risposta."""
    if not user_message or not session_id:
        raise HTTPException(status_code=400, detail="❌ I campi user_message e session_id sono obbligatori.")

    try:
        response = requests.post(
            f"{AI_BACKEND_URL}/chat",
            json={"user_message": user_message, "session_id": session_id},
            timeout=10  # ✅ Impostiamo un timeout massimo di 10 secondi
        )
        response.raise_for_status()  # ✅ Lancia un'eccezione per errori HTTP 4xx/5xx
        return response.json()
    except requests.exceptions.Timeout:
        logging.error("❌ Timeout nella comunicazione con il backend AI.")
        raise HTTPException(status_code=504, detail="❌ Timeout nella comunicazione con il backend AI.")
    except requests.exceptions.RequestException as e:
        logging.error(f"❌ Errore comunicazione AI: {str(e)}")
        raise HTTPException(status_code=500, detail=f"❌ Errore comunicazione AI: {str(e)}")

File: backend/routes/aiRoutes.js
const express = require("express");
const router = express.Router();
const axios = require("axios");
const rateLimit = require("express-rate-limit");
const winston = require("winston");

const AI_BACKEND_URL = "http://127.0.0.1:8000"; // URL locale del backend AI

// ✅ Configurazione del logging avanzato
const logger = winston.createLogger({
  level: "error",
  format: winston.format.json(),
  transports: [new winston.transports.File({ filename: "logs/ai_errors.log" })],
});

// ✅ Middleware per limitare le richieste API (Max 50 richieste per IP ogni 10 minuti)
const limiter = rateLimit({
  windowMs: 10 * 60 * 1000, // 10 minuti
  max: 50,
  message: "❌ Troppe richieste. Riprova più tardi.",
});

// ✅ Rotta per la chat con l'AI
router.post("/chat", limiter, async (req, res) => {
  try {
    const { user_message, session_id } = req.body;

    if (!user_message || !session_id) {
      return res
        .status(400)
        .json({
          error: "❌ I campi user_message e session_id sono obbligatori.",
        });
    }

    const response = await axios.post(
      `${AI_BACKEND_URL}/chat`,
      {
        user_message,
        session_id,
      },
      { timeout: 10000 }
    ); // ✅ Timeout di 10 secondi

    res.json(response.data);
  } catch (error) {
    if (error.response) {
      // Errore HTTP dal backend AI
      logger.error(
        `❌ Errore AI (${error.response.status}): ${error.response.data}`
      );
      return res.status(error.response.status).json({
        message: "❌ Errore nella comunicazione con l'AI",
        error: error.response.data,
      });
    } else if (error.code === "ECONNABORTED") {
      // Timeout nella richiesta
      logger.error("❌ Timeout nella comunicazione con l'AI.");
      return res
        .status(504)
        .json({ message: "❌ Timeout nella comunicazione con l'AI." });
    } else {
      // Errore generico
      logger.error(`❌ Errore AI: ${error.message}`);
      return res
        .status(500)
        .json({
          message: "❌ Errore nella comunicazione con l'AI",
          error: error.message,
        });
    }
  }
});

module.exports = router;
